egs/librispeech/ASR/pruned_transducer_stateless2/scaling.py

```diff
@@ -111,6 +111,74 @@ def backward(
         return x_grad - neg_delta_grad, None, None, None, None, None, None
 
 
+class GradientFilterFunction(torch.autograd.Function):
+    @staticmethod
+    def forward(
+        ctx,
+        x: Tensor,
+        batch_dim: int,  # e.g., 1
+        threshold: float,  # e.g., 10.0
+        *params: nn.Parameter,  # module parameters
+    ):
+        if x.requires_grad:
+            if batch_dim < 0:
+                batch_dim += x.ndim
+            ctx.batch_dim = batch_dim
+            ctx.threshold = threshold
+        return x, *params
+
+    @staticmethod
+    def backward(
+        ctx,
+        x_grad: Tensor,
+        *param_grads: nn.Parameter,
+    ):
+        dim = ctx.batch_dim
+        norm_dims = [d for d in range(x_grad.ndim) if d != dim]
+        norm_of_batch = (x_grad ** 2).mean(dim=norm_dims, keepdim=True).sqrt()
+        median_norm = norm_of_batch.median()
+        mask = norm_of_batch <= median_norm * ctx.threshold
+        x_grad = x_grad * mask
+
+        if not torch.all(mask):
+            # If any of elements of x_grad is zeroed,
+            # the param_grads would be fully zeroed.
+            for g in param_grads:
```

Comment:
in this case it is probably fine, but in general we are probably not supposed to use zero_() or other in-place operations, because the grad tensor could potentially be shared with some other tensor that will be used in future.  more safe to do: g = torch.zeros_like(g), for instance.

Reply:
I am afraid
```python3
g = torch.zeros_like(g)
```
will not change `param_grads` since it changes the temporary `g`.

Reply:
yes, would need different code construct, like param_grads = [somethng for x in param_grads]

Reply:
Finished it in the last commit https://github.com/k2-fsa/icefall/pull/564/commits/e71b541a21985e3ad2646f9215a7e9bda2536a91.